dask/base.py

```diff
@@ -132,14 +132,50 @@ def normalize_function(func):
 normalize_token.register(object,
         lambda a: normalize_function(a) if callable(a) else a)
 normalize_token.register(dict, lambda a: tuple(sorted(a.items())))
+
 with ignoring(ImportError):
     import pandas as pd
-    normalize_token.register(pd.DataFrame,
-            lambda a: (id(a), len(a), list(a.columns)))
-    normalize_token.register(pd.Series, lambda a: (id(a), len(a), a.name))
+
+    @partial(normalize_token.register, pd.Index)
+    def normalize_index(ind):
+        return [ind.name, normalize_token(ind.values)]
+
+    @partial(normalize_token.register, pd.Categorical)
+    def normalize_categorical(cat):
+        return [normalize_token(cat.codes),
+                normalize_token(cat.categories),
+                cat.ordered]
+
+    @partial(normalize_token.register, pd.Series)
+    def normalize_series(s):
+        return [s.name, s.dtype,
+                normalize_token(s._data.blocks[0].values),
+                normalize_token(s.index)]
+
+    @partial(normalize_token.register, pd.DataFrame)
+    def normalize_dataframe(df):
+        data = [block.values for block in df._data.blocks]
+        data += [df.columns, df.index]
+        return list(map(normalize_token, data))
+
+
 with ignoring(ImportError):
     import numpy as np
-    normalize_token.register(np.ndarray, lambda a: (id(a), a.dtype, a.shape))
+    @partial(normalize_token.register, np.ndarray)
+    def normalize_array(x):
+        if not x.shape:
+            return (str(x), x.dtype)
+        if x.dtype.hasobject:
+            try:
+                data = md5('-'.join(x.flat)).hexdigest()
+            except TypeError:
+                data = md5(b'-'.join([str(item).encode() for item in x.flat])).hexdigest()
+        else:
+            try:
+                data = md5(x.view('i1').data).hexdigest()
+            except (BufferError, AttributeError, ValueError):
+                data = md5(x.copy().view('i1').data).hexdigest()
```

Comment:
Sorry to necro an old thread. Just appears that the `i1` view has its origin here. Am curious why `u1` was not used for the view considering the intent?

Reply:
I suspect that we were trying to get a memoryview with a one-byte stride.  I don't currently recall exactly what broke without this.

Reply:
Since we're just trying to view the data as bytes it really doesn't matter whether it's signed or unsigned, just that the element is just a single byte in size. `u1` would have been an equally valid choice, but there's nothing wrong with `i1` - they give the same results:

```python
In [22]: import numpy as np

In [23]: from hashlib import md5

In [24]: a = np.array([-1, 1, 2])

In [25]: md5(a.view('i1').data).hexdigest()
Out[25]: '2c2020a2d6693493847c4d25749156e2'

In [26]: md5(a.view('u1').data).hexdigest()
Out[26]: '2c2020a2d6693493847c4d25749156e2'
```

Reply:
Yeah, that would be fine.  Was `u1` causing an issue?

Reply:
Guessing you meant `i1`, not that I know of. Just always found it a bit easier to follow unsigned bytes instead of signed bytes.